### ISSUE TITLE
Decorating with nested containers

### ISSUE THREAD

**Issue by iMoses** [young-steveo/bottlejs#71]
I'd like to decorate all providers within a nested container, problem is that the decorator callback returns the container itself instead of the service instance. I'm assuming this is a bug since it is inconsistent with the global behavior of not specifying a service name.

This works as expected according to the docs:
```
di.service(name, Class);
di.decorator(service => {
    console.log(service); // returns a Class instance
    return service;
});
```
But if we use nested containers:
```
di.service(`service.${name}`, Class);
di.decorator('service', service => {
    console.log(service); // returns a bottlejs container, expected to be a Class instance
    return service;
});
```

First, I wanna make sure this is indeed a bug and not a feature ;)
In case it is, I'll happily make a PR to fix this behavior.

**Comment by young-steveo** [comment]
@iMoses When you use dot notation, bottle creates a new container for the parent namespace and registers your services in that container, i.e.:

```js
bottle.service('Parent.Sub.Service', Thing);
bottle.container.Parent; // This is a container
bottle.container.Parent.Sub; // This is a container.
bottle.container.Parent.Sub.Service; // This is an instance of Thing
```
So in your case, you are assigning a decorator to the parent container:

```js
di.service(`service.${name}`, Class);
di.decorator('service', service => {
    console.log(service[name]); // this is the droid you're looking for
    return service;
]);
```

I would say that this is the expected behavior.  However, it sounds like what you want to do is define a decorator that will run on all children of a nested container.  That sounds like a reasonable feature I could add to the api of bottle.

To add this functionality, I'd have to either come up with a convention for decorator slugs (e.g. `"service.*"`), or write a new method (e.g. `bottle.decorateAll('service', ...)`.  I'm not sure which approach would be better, so I'll think about it for a bit.

**Comment by iMoses** [comment]
My personal opinion is to go with the asterisk, it follows known conventions and doesn't add anything to the interface.

**Comment by iMoses** [comment]
@young-steveo Any news about this feature?
I'll gladly help if help if needed.

**Comment by young-steveo** [comment]
@iMoses Thanks!  I'm going to start knocking out some of these "version 2" tickets this week.  For this feature I'm probably going to go with a new method on the sub container itself, rather than the asterisk version.  This will be very similar to the `$register` and `$list` functions that already exist.  For example, using the new method would look like this:

```js
di.service(`someService.${name}`, Class);
di.container.someService.$decorator(service => {
    // do stuff
    return service;
});
```

If you want to help out, I can start on the other tickets first.  The implementation should ensure that services are not instantiated until the "user" accesses them (lazy loading is a bottle feature) so it might be a little more complicated than accessing all of the sub keys in a loop and calling decorator on them (I think `Object.keys` will cause the "getters" to fire, but I'm not positive.)

**Comment by iMoses** [comment]
@young-steveo I'll give it a go this weekend with all that you said taken into account.
Do you have a guide for contributors or should I simply make a pull-request and verify that the feature is thoroughly tested?

**Comment by young-steveo** [comment]
Thanks, @iMoses!

There's a grunt task to lint and uglify the code.  Running `grunt test` will do do all that and run the jasmine tests.  The tests run against the minified versions in `/dist`.  Just don't commit the changed files in `/dist`, and everything should be good.

**Comment by young-steveo** [comment]
Oh, also keep in mind that bottle supports ECMAScript 5 compliant browsers, so the source can't use cool ES6 features like arrow functions.

**Comment by iMoses** [comment]
@young-steveo It actually seems to be pretty straight forward, I had to do very little to get it working and pass all tests.
Only test I can't pass is the "will allow decorators to be defined before services" because the `$decorator` method requires the container to be defined *before* you can add decorators to it. I'm assuming that it's an accepted behavior due to the nature of the feature.

**Comment by young-steveo** [comment]
This is awesome, thanks!  Since it looks backwards compatible, I think I'll make a `v1.6.0` branch instead of waiting to launch it with `v2.0.0` (assuming everything else I've merged into master is BC, which I think it is).

**Comment by young-steveo** [comment]
This is now available in [version 1.6.0](https://github.com/young-steveo/bottlejs/releases/tag/v1.6.0)
